### PR TITLE
add readyz handling to netexec

### PR DESCRIPTION
**Pull request by deads2k**
Testing services and endpoint handling requires readyz handling during termination of a pod.  Chasing a possible bug, I found this gap in the server where we have no external signal for a kubelet to mark ready=false.

/kind bug
/priority important-soon
/sig network

```release-note
NONE
```

### CODE REVIEW

File: test/images/agnhost/netexec/netexec.go [hunk 1]
```diff
@@ -213,14 +214,15 @@ func main(cmd *cobra.Command, args []string) {
...
-func addRoutes(mux *http.ServeMux, exitCh chan shutdownRequest) {
+func addRoutes(mux *http.ServeMux, termCh chan os.Signal, exitCh chan shutdownRequest) {
 	mux.HandleFunc("/", rootHandler)
 	mux.HandleFunc("/clientip", clientIPHandler)
 	mux.HandleFunc("/header", headerHandler)
 	mux.HandleFunc("/dial", dialHandler)
 	mux.HandleFunc("/echo", echoHandler)
 	mux.HandleFunc("/exit", func(w http.ResponseWriter, req *http.Request) { exitHandler(w, req, exitCh) })
 	mux.HandleFunc("/healthz", healthzHandler)
+	mux.HandleFunc("/readyz", readyzHandler(termCh))
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:225]: can you add the description of the new handler? , L93 per example
https://github.com/kubernetes/kubernetes/pull/110174/files#diff-f26ab21963f05c932bece947c19bcff504d1f6a4d53e4e33f2029908126bd12aR93

File: test/images/agnhost/netexec/netexec.go [hunk 2]
```diff
@@ -339,6 +341,32 @@ func healthzHandler(w http.ResponseWriter, r *http.Request) {
...
+		select {
+		case <-termCh:
+			w.WriteHeader(http.StatusServiceUnavailable)
+			if _, err := w.Write([]byte("shutting down")); err != nil {
+				utilruntime.HandleError(err)
+			}
+			return
+
+		default:
+			if serverReady.get() {
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:359]: if we set the `serverReady.get()` variable as unready based on the term signal, will that not work for both healthz and readyz handler and simplify the code here?

**Comment by aojea** [comment]
/assign @aojea @andrewsykim 

hmm, this handle was added to being able to test the "termination endpoints" functionality IIRC

https://github.com/kubernetes/kubernetes/pull/108750

I need to check carefully if this is going to break it

The problem is that the changes in the images are not tested directly in the CI, the images have to be promoted and the k/k code updated with the new image, so the CI picks it up ... there is a trick to build the image locally and publish in a personal registry and add a temporary commit to test the new image in the CI

/hold

**Comment by aojea** [comment]
There are more changes needed for the image, like updating the version

https://github.com/kubernetes/kubernetes/blob/5ee7de7b5cd682380e036dc769cd1fde053a6626/test/images/agnhost/VERSION

the whole process is documented here

https://github.com/kubernetes/kubernetes/tree/master/test/images#updating-the-tests-images

a trick for testing your changes in the CI is to build your custom image and publish to a personal repo, as example

```sh
$ cd test/images
$ REGISTRY=quay.io/aojea make all-push WHAT=agnhost
```

and then "hijack" the image by replacing it in the manifest used to configure the e2e registries

https://github.com/kubernetes/kubernetes/blob/aa49dffc7f24dc31dd333869be8e6e3cdfc00af9/test/utils/image/manifest.go#L236

or just adding the image url directly to the tests

**Comment by deads2k** [comment]
updated for version and comments.

**Comment by deads2k** [comment]
attempting proof in https://github.com/kubernetes/kubernetes/pull/110176

**Comment by aojea** [comment]
/retest
/lgtm

Tests with the new image pass
https://github.com/kubernetes/kubernetes/pull/110176

@deads2k feel free to unhold, remember you still need 2 additional steps https://github.com/kubernetes/kubernetes/tree/master/test/images#promoting-images

1. wait for the postubmit job and PR k8s.io to promote the image
2. update k/k tests with the new image

**Comment by deads2k** [comment]
> @deads2k feel free to unhold, remember you still need 2 additional steps https://github.com/kubernetes/kubernetes/tree/master/test/images#promoting-images
> 
> 1. wait for the postubmit job and PR k8s.io to promote the image
> 2. update k/k tests with the new image

Acknolwedged

/hold cancel

File: test/images/agnhost/netexec/netexec.go [hunk 3]
```diff
@@ -161,11 +166,18 @@ func (a *atomicBool) get() bool {
...
+	sigTermReceived := make(chan struct{})
+	go func() {
 		termCh := make(chan os.Signal, 1)
 		signal.Notify(termCh, syscall.SIGTERM)
+
+		<-termCh
+		close(sigTermReceived)
+	}()
+
+	if delayShutdown > 0 {
```
Comment (aojea) [inline test/images/agnhost/netexec/netexec.go:178]: this is the problem, if not delayShutdown is set the process keeps hanging on sigterm

Comment (deads2k) [reply test/images/agnhost/netexec/netexec.go:178]: > this is the problem, if not delayShutdown is set the process keeps hanging on sigterm

Which particular line does the process hang on?  I don't mind reverting or fixing, but this line existed gating off this go func before.

Comment (aojea) [reply test/images/agnhost/netexec/netexec.go:178]: before, we didn't capture the sigterm signal if delayShutdown was 0.
Now, we always capture the sigterm signal, but if we have delayShutdown == 0, we don't execute the os.Exit() and we don't signal the process to exit

**Pull request merged**